src/peer_connection.cpp

```diff
@@ -6518,8 +6518,9 @@ namespace libtorrent
 		peer_log(peer_log_alert::outgoing_message, "KEEPALIVE");
 #endif
 
-		m_last_sent = aux::time_now();
-		write_keepalive();
+		if (write_keepalive()) {
+			m_last_sent = aux::time_now();
```

Comment:
yeah, I think either way is fine. If the connection truly is stalled, setup_send() won't be called either, so in practice I don't think it would make a significant difference